### PR TITLE
FlatTheme support

### PR DESCRIPTION
**Pull request by FraukeF**
Thank you for contributing to harp.gl!

Before requesting a pull request, please remember to check the following documents:
* [contribution guidelines](https://github.com/heremaps/harp.gl/blob/master/CONTRIBUTING.md)
* [coding style](https://github.com/heremaps/harp.gl/blob/master/CODINGSTYLE.md)

If you are adding new functionality we would highly appreciate if you can describe what is the capability you are adding and even better if you can add some examples. Please also remember to add tests for it.

# CI Check

Our bots will check whether your PR can be directly integrated into the mainline. We have some internal integration tests running on the background, our bots will inform you of the next steps and someone from our team will take a look and help if needed!

And please do not forget to sign-off your commit! You can read more about DCO [here](https://julien.ponge.org/blog/developer-certificate-of-origin-versus-contributor-license-agreements/). But, in short, you just need to use `git commit -s` or append `--signoff` when you are committing to the repo.

Happy contributing!

### CODE REVIEW

File: @here/harp-datasource-protocol/lib/Theme.ts [hunk 1]
```diff
@@ -252,6 +252,33 @@ export function isJsonExprReference(value: any): value is JsonExprReference {
...
 
+/**
+ * Extracts a StyleSet for a styleSetName from a Theme
+ *
+ * @param theme - The theme
+ * @param styleSetName - The name of the styleSet to be extracted.
+ *
+ * @return The Extracted StyleSet
+ */
+export function extractStyleSet(theme: Theme | FlatTheme, styleSetName: string): StyleSet {
```
Comment (robertoraggi) [inline @here/harp-datasource-protocol/lib/Theme.ts:263]: please mark this function with `@internal`

File: @here/harp-mapview-decoder/lib/TileDataSource.ts [hunk 2]
```diff
@@ -198,15 +205,15 @@ export class TileDataSource<TileType extends Tile = Tile> extends DataSource {
...
         }
 
-        const styleSet =
-            this.styleSetName !== undefined && theme.styles
-                ? theme.styles[this.styleSetName]
-                : undefined;
+        let styleSet: StyleSet | undefined;
+        if (this.styleSetName !== undefined) {
+            styleSet = extractStyleSet(theme, this.styleSetName);
+        }
```
Comment (robertoraggi) [inline @here/harp-mapview-decoder/lib/TileDataSource.ts:216]: filtering the style rules here should be deprecated and considered only as a temporary solution until we are with the deprecation of the old theme format.

Our goal should be to deprecate (and eventually remove) the concept that there is only one single "styleSet" associated to a datasource. That is, currently the datasources must set the property `"styleSetName"` at construction type and then later this property is used to filter the style rules. This is not how the unified theme is supposed to work.

In the unified theme format the `styleSet` is just a way to group together style rules, for example a styleSet called "roads" could be used to group layers for the roads, another one called city-center could collect the layers for the city centers and so on.. These groups of rules could later be references in `theme.priorities` and `theme.labelPriorities` to source items. Eventually we could also have APIs to enable/disable these style sets.

File: @here/harp-mapview-decoder/lib/TileDataSource.ts [hunk 3]
```diff
@@ -198,15 +205,15 @@ export class TileDataSource<TileType extends Tile = Tile> extends DataSource {
      * `styleSetName` property) is found in `theme`.
      * @override
      */
-    setTheme(theme: Theme, languages?: string[], styleSetName?: string): void {
+    setTheme(theme: Theme | FlatTheme, languages?: string[], styleSetName?: string): void {
```
Comment (robertoraggi) [inline @here/harp-mapview-decoder/lib/TileDataSource.ts:208]: i'm not sure that the signature of this function is correct. This code above is assuming that the `theme` set to this data source is already fully resolved (e.g. all the extends are evaluated), which is not the case.
I would expect setTheme to be an `async` function, accept a url or a theme and resolve the theme using the ThemeLoader.

Comment (robertoraggi) [reply @here/harp-mapview-decoder/lib/TileDataSource.ts:208]: also i would remove `styleSetName` and languages from the signature of this function.

Comment (FraukeF) [reply @here/harp-mapview-decoder/lib/TileDataSource.ts:208]: Will be added that in #1949

File: @here/harp-datasource-protocol/lib/Theme.ts (hunk 1 above)
Comment (FraukeF) [reply @here/harp-datasource-protocol/lib/Theme.ts]: removed it

File: @here/harp-mapview-decoder/lib/TileDataSource.ts (hunk 2 above)
Comment (FraukeF) [reply @here/harp-mapview-decoder/lib/TileDataSource.ts]: I added your comment to the ticket I created for the full FlatTheme support: HARP-12835

File: @here/harp-mapview-decoder/lib/TileDataSource.ts (hunk 3 above)
Comment (FraukeF) [reply @here/harp-mapview-decoder/lib/TileDataSource.ts]: I squashed the two PRs, so the resolving of the themes is done... as well as the removal of the styleSetName I had added...

**Review by robertoraggi — APPROVED** [review]

**Pull request merged**